lemur/plugins/lemur_google_ca/plugin.py

```diff
@@ -0,0 +1,214 @@
+"""
+.. module: lemur.plugins.lemur_google_ca.plugin
+    :platform: Unix
+    :synopsis: This module is responsible for creating certificates with the Google CA API '
+    :license: Apache, see LICENSE for more details.
+
+    Google CA (v2 API) Documentation
+    https://cloud.google.com/certificate-authority-service/docs/reference/rest
+
+    This plugin requires following packages:
+    - google-cloud-private-ca
+    - protobuf
+    - types-protobuf (for mypy)
+    Make sure to add these to `requirements.in`
+
+    The plugin requires `GOOGLE_ACCOUNT_CREDENTIALS` config variable, which should point at the file containing
+    credentials that Lemur is using to connect to Google Cloud Platform.
+
+    IAM permissions:
+    To issue a certificate, Lemur would need permission `privateca.certificates.create`
+    for the specified Certifiate authority
+
+    To revoke a certificate, Lemur would need permission `privateca.certificates.update`
+    for the specified Certifiate authority
+
+    To add a Google-based CA, Lemur would need permission `privateca.certificateAuthorities.get`
+
+    This can be achieved by assigning `roles/privateca.certificateAuthorityViewer` and `
+    roles/privateca.certificateManager` to Lemur's service account, or by using a custom role.
+
+.. moduleauthor:: Oleg Dopertchouk <odopertchouk@squarespace.com>
+"""
+import json
+import re
+import uuid
+from typing import Optional
+
+import google.cloud.security.privateca_v1 as private_ca
+
+import arrow
+from flask import current_app
+from google.oauth2 import service_account
+from google.protobuf import duration_pb2
+
+from lemur.constants import CRLReason
+from lemur.common.utils import validate_conf
+import lemur.plugins.lemur_google_ca
+from lemur.plugins.bases import IssuerPlugin
+
+import google.cloud.security.privateca_v1 as privateca_v1
+
+SECONDS_PER_YEAR = 365 * 24 * 60 * 60
+
+
+def get_duration(options):
+    """
+    Deduce certificate duration from options
+    """
+    validity_end = options.get("validity_end")
+    if validity_end:
+        return int((validity_end - arrow.utcnow()).total_seconds())
+    else:
+        return options.get("validity_years", 1) * SECONDS_PER_YEAR
+
+
+def generate_certificate_id(common_name) -> str:
+    """
+    Generates a readable unique id for a cert based on cert's CN
+    """
+    name = common_name.lower().strip()
+    name = re.sub(r'[^a-z0-9-]', '_', name)
+    name = name[:50]  # leave space for random id
+    return f"{name}-{uuid.uuid4().hex}"[:63]  # Truncate to 63 characters, to fit the api constraints
+
+
+def fetch_authority(ca_path: str) -> tuple[str, str]:
+    client = privateca_v1.CertificateAuthorityServiceClient()
```

Comment:
Does this need application credentials as well?

Reply:
Good catch!